app/models/ems_refresh/refreshers/ems_refresher_mixin.rb

```diff
@@ -52,8 +45,76 @@ def refresh
         _log.info "Refreshing all targets...Complete"
       end
 
-      def save_inventory(ems, _targets, hashes)
-        EmsRefresh.save_ems_inventory(ems, hashes)
+      def preprocess_targets
+        @full_refresh_threshold = options[:full_refresh_threshold] || 10
+
+        # See if any should be escalated to a full refresh
+        @targets_by_ems_id.each do |ems_id, targets|
+          ems = @ems_by_ems_id[ems_id]
+          ems_in_list = targets.any? { |t| t.kind_of?(ExtManagementSystem) }
+
+          if ems_in_list
+            _log.info "Defaulting to full refresh for EMS: [#{ems.name}], id: [#{ems.id}]." if targets.length > 1
+            targets.clear << ems
+          elsif !ems_in_list && targets.length >= @full_refresh_threshold
+            _log.info "Escalating to full refresh for EMS: [#{ems.name}], id: [#{ems.id}]."
+            targets.clear << ems
+          end
+        end
+      end
+
+      def refresh_targets_for_ems(ems, targets)
+        # handle a 3-part inventory refresh process
+        # 1. collect inventory
+        # 2. parse inventory
+        # 3. save inventory
+        log_header = format_ems_for_logging(ems)
+
+        targets_with_inventory, _ = Benchmark.realtime_block(:get_inventory_for_targets) do
+          get_inventory_for_targets(ems, targets)
+        end
+
+        until targets_with_inventory.empty?
+          target, inventory = targets_with_inventory.shift
+
+          _log.info "#{log_header} Refreshing target #{target.class} [#{target.name}] id [#{target.id}]..."
+          hashes, _ = Benchmark.realtime_block(:parse_targeted_inventory) do
+            parse_targeted_inventory(ems, inventory, target)
+          end
+          inventory = nil # clear to help GC
+
+          Benchmark.realtime_block(:save_inventory) { save_inventory(ems, target, hashes) }
+          _log.info "#{log_header} Refreshing target #{target.class} [#{target.name}] id [#{target.id}]...Complete"
+        end
+      end
+
+      def get_inventory_for_targets(ems, targets)
+        # TODO: implement this method in all refreshers and remove from here
+        # legacy refreshers collect inventory during the parse phase so the
+        # inventory component of the return value is empty
+        #
+        # override this method and return an array of:
+        #   [[target1, inventory_for_target1], [target2, inventory_for_target2]]
+        targets.collect { |target| [target, []] }
```

Comment:
Is this base implementation for the legacy stuff?  If so, maybe it would be nicer to have a private method called `legacy_get_inventory_for_targets`, which would be called by `get_inventory_for_targets`.  That way, it's clear that this is an implementation for legacy.